### PR TITLE
134448159 structured data markup breadcrumb

### PR DESCRIPTION
**Pull request by cesswairimu**
Addresses: https://www.pivotaltracker.com/story/show/134448159

### CODE REVIEW

**Comment by tansaku** [comment]
well done getting started on this @cesswairimu 

does the build failure and codeclimate comments make sense?

**Comment by cesswairimu** [comment]
@tansaku no i do not know why some cucumber tests are failing

**Comment by tansaku** [comment]
hmm, odd failure:

```
  1) Organisation find all orgs that have keyword anywhere in their name or description
     Failure/Error: expect(Organisation.search_by_keyword('elderly')).to eq([@org2, @org3])
     
       expected: [#<Organisation id: 304, name: "Indian Elders Association", address: "64 pinner road", postcode: "HA1...ublish_email: true, deleted_at: nil, type: "Organisation", non_profit: nil, slug: "age-uk-elderly">]
            got: #<ActiveRecord::Relation [#<Organisation id: 305, name: "Age UK Elderly", address: "64 pinner road", ...: true, deleted_at: nil, type: "Organisation", non_profit: nil, slug: "indian-elders-association">]>
     
       (compared using ==)
     
rspec ./spec/models/organisation_spec.rb:206 # Organisation find all orgs that have keyword anywhere in their name or description
```

**Comment by tansaku** [comment]
it passes locally for me - how about you?

I've restarted the CI build to see if it recurs ...

**Comment by tansaku** [comment]
also, what do you think of the codeclimate issues? https://codeclimate.com/github/AgileVentures/LocalSupport/pull/429

**Comment by tansaku** [comment]
ah, that rspec fail looks like a random thing (not good, and maybe worthy of investigation later) however now we see some cukes fail:

```
Failing Scenarios:
cucumber -p second_try features/admin/admin_new_charity.feature:81 # Scenario: Successfully create charity while being signed-in as superadmin from arbitrary page
cucumber -p second_try features/home_page/home_button.feature:27 # Scenario Outline: Find help with care for elderly, Examples (#1)
cucumber -p second_try features/home_page/home_button.feature:28 
```

and these I can replicate locally, where I get the following message:

```
  Scenario Outline: Find help with care for elderly         # features/home_page/home_button.feature:20
    Given I visit the <page>                                # features/home_page/home_button.feature:21
    Then I should see an <status> home button in the header # features/home_page/home_button.feature:22
    And I click "Home"                                      # features/home_page/home_button.feature:23
    Then I should be on the home page                       # features/home_page/home_button.feature:24

    Examples: 
      | page                                          | status   |
      | volunteer opportunities page                  | inactive |
      Ambiguous match, found 2 elements matching link "Home" (Capybara::Ambiguous)
      ./features/step_definitions/navigation_steps.rb:130:in `/^I click "(.*)"$/'
      features/home_page/home_button.feature:28:in `And I click "Home"'
      features/home_page/home_button.feature:23:in `And I click "Home"'

Failing Scenarios:
cucumber features/home_page/home_button.feature:28 # Scenario Outline: Find help with care for elderly, Examples (#2)
```

does that error message give you an idea about what might be wrong?

**Comment by cesswairimu** [comment]
yeah @tansaku  locally rspec is okay but cucumber has two failures. About code climate i think its about the " " and i have changed them to single quotes

**Comment by tansaku** [comment]
this is looking great @cesswairimu - you're really nailing these now :-)

So to go to the next level I wonder if we can connect this PR up so that it has a high level description and connects to the PivotalTracker ticket (so that gets auto-closed) when I merge this?

There's a description of the kind of commit message needed here: https://github.com/AgileVentures/LocalSupport/blob/develop/CONTRIBUTING.md#git-comments

And the sort of PR description we would like here:
https://github.com/AgileVentures/LocalSupport/blob/develop/CONTRIBUTING.md#pull-requests

**Comment by cesswairimu** [comment]
@tansaku yes u think that would be brilliant

**Comment by tansaku** [comment]
@cesswairimu looking good :-)



note that we can make the current page a hyperlink too 

BTW, did my [suggestions](https://github.com/AgileVentures/LocalSupport/pull/429#issuecomment-285317687) about the PR description and commit messages make sense?  Do you need help adjusting the current PR description?

**Comment by tansaku** [comment]
regarding the code climate issues we could break the breadcrumb statements out to separate methods from:

```rb
  def show
    render template: 'pages/404', status: 404 and return if @volunteer_op.nil?
    @organisation = Organisation.friendly.find(@volunteer_op.organisation_id)
    organisations = Organisation.where(id: @organisation.id)
    @editable = current_user.can_edit?(@organisation) if current_user
    @markers = BuildMarkersWithInfoWindow.with(VolunteerOp.build_by_coordinates, self)
    add_breadcrumb @organisation.name, organisation_path(@organisation)
    add_breadcrumb @volunteer_op.title, :volunteer_op_path
  end
```

perhaps I can show you how in today's meeting ...?

File: features/step_definitions/navigation_steps.rb [hunk 1]
```diff
@@ -130,6 +130,17 @@ def find_record_for(object, schema, name)
   click_link(link)
 end
 
+When /^I click "(.*)" (.*) link$/ do |link, position|
```
Comment (sigu) [inline features/step_definitions/navigation_steps.rb]: @tansaku , I had a feeling we could have improved on the previous step regular expression to accommodate this little change introduced
``` cucumber
 When /^I click "(.*)"$/ do |link|
   click_link(link)
 end
```
Looks like having both is not DRY enough

**Comment by cesswairimu** [comment]
@tansaku yes i understood the steps in contributing. I will try implement the hyperlinks in the last breadcrumb trail. If i succeed i will make the last commit  [Finishes #ticket_id]. Thanks for that

File: features/step_definitions/navigation_steps.rb (hunk 1 above)
Comment (tansaku) [inline features/step_definitions/navigation_steps.rb]: maybe we should be using `within` to isolate the clicking on the breadcrumb via it's id?

**Comment by cesswairimu** [comment]
@tansaku  @sigu The gem i was using for this ticket 'breadcrumbs_on_rails' is set in a way that the last link on the trail is not clickable. Trying to make the last element clickable was unsuccessful.

**Comment by tansaku** [comment]
@cesswairimu that's fine - Amazon doesn't make them clickable - let's leave it as it is :-)

**Comment by sigu** [comment]
@tansaku this looks ready to me for merging. Have a final look and possibly merge it in

**Comment by tansaku** [comment]
great stuff @cesswairimu and @sigu - could you make refactoring tickets for each code climate issue so that we have a record of them when I merge?

**Comment by tansaku** [comment]
finally - could we get a PT ticket closing commit? https://github.com/AgileVentures/LocalSupport/blob/develop/CONTRIBUTING.md#git-comments

**Comment by cesswairimu** [comment]
Structured data markup breadcrumb trail for easy navigation [Finishes #134448159]

**Comment by tansaku** [comment]
great @cesswairimu - if you could just finally paste in links to refactoring tickets in pivotal tracker corresponding to the code climate issues that aren't being fixed, then we'll be all done :-)

**Comment by cesswairimu** [comment]
ticket for refactoring the code climate issues https://www.pivotaltracker.com/n/projects/742821/stories/144806617

**Pull request merged**